### PR TITLE
fix tcp launcher/tailer deadlock on agent shutdown

### PR DESCRIPTION
**Pull request by gh123man**
### What does this PR do?

When the agent shut down the launcher's `Stop()` function gets called. This first closes the listener and then shuts down each tailer. However when the listener is closed `tailer.Conn.Read(frame)` returns an error (on a different goroutine) causing the tailer to be stopped. There are 2 problems:

1. the tailer can only be stopped once otherwise it will block. So this change makes sure `Stop` can only be called once on a tailer from by the launcher ensuring it's removed from the active tailer list in both stop cases. 
2. there was a race between the 2 stop cases where if the stop from the connection being closed won the race (background stop), the `launcher.Stop` function would deadlock because it was stopped previously (foreground stop). If the race was lost, we would leak a goroutine since the backround `Stop` would deadlock, but not block a critical thread. 

### Motivation

Rare test failure

### Additional Notes

This is difficult to validate. I ran `TestTCPShouldReceivesMessages` in a loop `8000` times to reproduce and validate the fix. I don't believe it's appropriate to leave such a unit test in our code. 

### Describe how to test/QA your changes

This only affected shutdown. 
- Start the agent with the TCP listener enabled and make sure it shuts down gracefully. 
- make sure the TCP listener still works correctly. 

### Reviewer's Checklist

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

File: releasenotes/notes/fix-tcp-listener-deadlock-d441552a6e64263e.yaml [hunk 1]
```diff
@@ -0,0 +1,11 @@
...
+# CHANGELOG.rst is rendered. So the text needs to be worded so that
+# it does not depend on any information only available in another
+# section. This may mean repeating some details, but each section
+# must be readable independently of the other.
+#
+# Each section note must be formatted as reStructuredText.
+---
+fixes:
+  - |
+    Fix potential deadlock on agent shutdown in TCP listener.
```
Comment (apigirl) [inline releasenotes/notes/fix-tcp-listener-deadlock-d441552a6e64263e.yaml:11]:
```suggestion
    Fix potential deadlock on Agent shutdown in TCP listener. 
```

**Review by apigirl — COMMENTED** [review]
one nit and then looks good for docs

**Review by djmitche — APPROVED** [review]
There seem to be a lot of races around shutting down the logs-agent.  Nice to see one fewer!

File: releasenotes/notes/fix-tcp-listener-deadlock-d441552a6e64263e.yaml [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+# CHANGELOG.rst is rendered. So the text needs to be worded so that
+# it does not depend on any information only available in another
+# section. This may mean repeating some details, but each section
+# must be readable independently of the other.
+#
+# Each section note must be formatted as reStructuredText.
+---
+fixes:
+  - |
+    Fix potential deadlock on Agent shutdown in TCP listener.
```
Comment (hush-hush) [inline releasenotes/notes/fix-tcp-listener-deadlock-d441552a6e64263e.yaml:11]: We want to give a bit more context to users. Like mentioning it's about the logs Agent.

WDYT:

```suggestion
    Fix potential deadlock when shutting down an Agent with a log TCP listener.
```

**Review by hush-hush — APPROVED** [review]

**Pull request merged**